### PR TITLE
Add AWS Signature Version 4

### PR DESCRIPTION
**Pull request by simov**
Closes https://github.com/request/request/pull/2031

- [AWS Signature Version 4](http://docs.aws.amazon.com/general/latest/gr/signature-version-4.html
https://www.npmjs.com/package/aws4)
- [aws4](https://www.npmjs.com/package/aws4) module

### CODE REVIEW

**Comment by mirkods** [comment]
When this will be merged?

**Comment by simov** [comment]
I'm thinking about moving the require statement down below inside the aws handler and make aws4 external dependency. I don't feel comfortable adding more dependencies in request. Similar to how it's done in [request/core](https://github.com/request/core).

**Comment by mirkods** [comment]
Ok feel free to apply your edits, you know what are a good stuff for the module

**Comment by simov** [comment]
2.68 is published :tada:

File: request.js [hunk 1]
```diff
@@ -1229,29 +1229,52 @@ Request.prototype.aws = function (opts, now) {
...
-    { key: opts.key
-    , secret: opts.secret
-    , verb: self.method.toUpperCase()
-    , date: date
-    , contentType: self.getHeader('content-type') || ''
-    , md5: self.getHeader('content-md5') || ''
-    , amazonHeaders: aws.canonicalizeHeaders(self.headers)
+  
+  if (opts.sign_version == 4 || opts.sign_version == '4') {
+    var aws4 = require('aws4')
```
Comment (rmg) [inline request.js]: This dependency was only added as a devDependency. See #2040

**Comment by rmg** [comment]
Sorry for the spam from commenting in multiple places, but this was a breaking change released as a patch. See #2040 and #2041.

**Pull request merged**